File: 2020/rust/src/day12.rs
use std::time::{Duration, Instant};

#[derive(Debug)]
struct Instruction {
    pub direction: Direction,
    pub amount: i64,
}

#[derive(Debug, PartialEq, Eq)]
enum Direction {
    North,
    South,
    East,
    West,
    Left,
    Right,
    Forward,
}

fn parse_input(input: &str) -> (Vec<Instruction>, Duration) {
    let start = Instant::now();
    let out = input
        .lines()
        .map(|l| l.as_bytes())
        .map(|line| {
            let direction = match line[0] {
                b'N' => Direction::North,
                b'S' => Direction::South,
                b'E' => Direction::East,
                b'W' => Direction::West,
                b'R' => Direction::Right,
                b'L' => Direction::Left,
                b'F' => Direction::Forward,
                _ => unreachable!("invalid input"),
            };
            let amount = unsafe { std::str::from_utf8_unchecked(&line[1..]) }
                .parse()
                .unwrap();

            Instruction { direction, amount }
        })
        .collect();
    let end = Instant::now();
    (out, end - start)
}

fn manhattan_distance(x1: i64, y1: i64, x2: i64, y2: i64) -> i64 {
    (x1 - x2).abs() + (y1 - y2).abs()
}

fn solve_part1(input: &[Instruction]) -> (i64, Duration) {
    let start = Instant::now();
    let mut x = 0;
    let mut y = 0;
    let mut direction = Direction::East;

    for instruction in input {
        match instruction {
            Instruction {
                direction: Direction::North,
                amount,
            } => {
                y += amount;
            }
            Instruction {
                direction: Direction::South,
                amount,
            } => {
                y -= amount;
            }
            Instruction {
                direction: Direction::East,
                amount,
            } => {
                x += amount;
            }
            Instruction {
                direction: Direction::West,
                amount,
            } => {
                x -= amount;
            }
            Instruction {
                direction: Direction::Right,
                amount,
            } => {
                for _ in 0..(amount / 90) {
                    direction = match direction {
                        Direction::North => Direction::East,
                        Direction::East => Direction::South,
                        Direction::South => Direction::West,
                        Direction::West => Direction::North,
                        _ => unreachable!(),
                    }
                }
            }
            Instruction {
                direction: Direction::Left,
                amount,
            } => {
                for _ in 0..(amount / 90) {
                    direction = match direction {
                        Direction::North => Direction::West,
                        Direction::East => Direction::North,
                        Direction::South => Direction::East,
                        Direction::West => Direction::South,
                        _ => unreachable!(),
                    }
                }
            }
            Instruction {
                direction: Direction::Forward,
                amount,
            } => match direction {
                Direction::North => y += amount,
                Direction::South => y -= amount,
                Direction::East => x += amount,
                Direction::West => x -= amount,
                _ => unreachable!(),
            },
        }
    }

    let out = manhattan_distance(0, 0, x, y);

    let end = Instant::now();
    (out, end - start)
}

fn solve_part2(input: &[Instruction]) -> (i64, Duration) {
    let start = Instant::now();
    let mut ship_x = 0;
    let mut ship_y = 0;
    let mut waypoint_x = 10;
    let mut waypoint_y = 1;

    for instruction in input {
        match instruction {
            Instruction {
                direction: Direction::North,
                amount,
            } => waypoint_y += amount,
            Instruction {
                direction: Direction::South,
                amount,
            } => waypoint_y -= amount,
            Instruction {
                direction: Direction::East,
                amount,
            } => waypoint_x += amount,
            Instruction {
                direction: Direction::West,
                amount,
            } => waypoint_x -= amount,
            Instruction {
                direction: Direction::Right,
                amount,
            } => {
                for _ in 0..(amount / 90) {
                    let tmp = waypoint_y;
                    waypoint_y = -waypoint_x;
                    waypoint_x = tmp;
                }
            }
            Instruction {
                direction: Direction::Left,
                amount,
            } => {
                for _ in 0..(amount / 90) {
                    let tmp = waypoint_x;
                    waypoint_x = -waypoint_y;
                    waypoint_y = tmp;
                }
            }
            Instruction {
                direction: Direction::Forward,
                amount,
            } => {
                ship_x += waypoint_x * amount;
                ship_y += waypoint_y * amount;
            }
        }
    }

    let out = manhattan_distance(0, 0, ship_x, ship_y);

    let end = Instant::now();
    (out, end - start)
}

pub fn run(input: String) {
    let (parsed_input, time) = parse_input(&input);
    println!("Day 12, parsing input took {}ns", time.as_nanos());
    let (p1, time) = solve_part1(&parsed_input);
    println!("Day 12, part 1: {}. Took {}ns", p1, time.as_nanos());
    let (p2, time) = solve_part2(&parsed_input);
    println!("Day 12, part 2: {}. Took {}ns", p2, time.as_nanos());
}

File: 2020/cpp/day_12.cpp
#include "day_12.h"

day_12::instruction_t::instruction_t(char opcode, ulong operand) :
    opcode {opcode},
    operand {operand} {}

day_12::day_12() :
    input {[] {
        const auto input_strings = readFile("Inputs/day_12.txt");
        const auto start = std::chrono::high_resolution_clock::now();
        std::vector<instruction_t> input;
        input.reserve(input_strings.size());
        for (const auto& line : input_strings) {
            input.emplace_back(
                *line.data(),
                static_cast<int>(fast_atol(line.data() + 1)));
        }
        const auto end = std::chrono::high_resolution_clock::now();
        fmt::print(
            "Parsing input for day twelve took {}ns\n",
            std::chrono::duration_cast<std::chrono::nanoseconds>(end - start)
                .count());
        return input;
    }()} {}

void day_12::part_one() const {
    const auto start = std::chrono::high_resolution_clock::now();
    int x {0};
    int y {0};
    int direction {1};
    for (auto [opcode, operand] : input) {
        switch (opcode) {
            case 'R':
                direction += operand / 90;
                break;
            case 'L':
                direction -= operand / 90;
                break;
            case 'F':
                switch (direction % 4) {
                    case 0:  // north
                        y += operand;
                        break;
                    case 1:  // east
                        x += operand;
                        break;
                    case 2:  // south
                        y -= operand;
                        break;
                    case 3:  // west
                        x -= operand;
                        break;
                }
                break;
            case 'N':
                y += operand;
                break;
            case 'E':
                x += operand;
                break;
            case 'S':
                y -= operand;
                break;
            case 'W':
                x -= operand;
                break;
        }
    }
    const auto distance = manhattan_distance(x, y, 0, 0);
    auto end = std::chrono::high_resolution_clock::now();

    fmt::print("The answer for day twelve part one is {}\n", distance);
    fmt::print(
        "Took {}ns\n",
        std::chrono::duration_cast<std::chrono::nanoseconds>(end - start)
            .count());
}

void day_12::part_two() const {
    const auto start = std::chrono::high_resolution_clock::now();
    int x {0};
    int y {0};
    int waypoint_x {10};
    int waypoint_y {1};
    for (auto [opcode, operand] : input) {
        switch (opcode) {
            case 'N':
                waypoint_y += operand;
                break;
            case 'E':
                waypoint_x += operand;
                break;
            case 'S':
                waypoint_y -= operand;
                break;
            case 'W':
                waypoint_x -= operand;
                break;
            case 'L':
                for (int i = 0; i < static_cast<int>(operand / 90); ++i) {
                    auto tmp = waypoint_x;
                    waypoint_x = -waypoint_y;
                    waypoint_y = tmp;
                }
                break;
            case 'R':
                for (int i = 0; i < static_cast<int>(operand / 90); ++i) {
                    auto tmp = waypoint_y;
                    waypoint_y = -waypoint_x;
                    waypoint_x = tmp;
                }
                break;
            case 'F':
                y += waypoint_y * operand;
                x += waypoint_x * operand;
                break;
        }
    }

    const auto distance = manhattan_distance(x, y, 0, 0);

    auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day twelve part two is {}\n", distance);
    fmt::print(
        "Took {}ns\n",
        std::chrono::duration_cast<std::chrono::nanoseconds>(end - start)
            .count());
}
